### ISSUE TITLE
Decorator changes signature of function that uses `TypeVar`

### ISSUE THREAD

**Issue by wch** [microsoft/pyright#3344]
This is a bit hard to describe, so I'll illustrate with an example.

`f` is a function which takes two parameters, of types `T` and `Union[T, List[T]]`. When we use `f()` and pass in those different types, the type checker correctly says that it's OK (no red underlining):

```py
from typing import Callable, TypeVar, Union, List, Any

T = TypeVar("T", int, str)

def f(x: T, y: Union[T, List[T]]) -> None:
    pass

f(1, 2)      # OK
f(1, [2, 3]) # OK
```

Now, let's say we have a decorator which returns a function of type `Callable[[U], U]`. (Imagine, for example, that this decorator is used to modify a docstring.) If that decorator is used on a function identical to `f`, then Pylance thinks that the signature of the resulting function has changed, and flags spurious errors.

```py
U = TypeVar("U")

def doc() -> Callable[[U], U]:
    ...

@doc()
def g(x: T, y: Union[T, List[T]]) -> None:
    pass

g(1, 2)      # `2` is underlined
g(1, [2, 3]) # OK
```

Here's a screenshot:
<img width="508" alt="image" src="https://user-images.githubusercontent.com/86978/163421231-495e446b-79a7-431e-8251-f41eb3b2512d.png">

When I hover over the `2`, I see this text:

```
Argument of type "Literal[2]" cannot be assigned to parameter "y" of type "List[T@g]" in function "g"
  "Literal[2]" is incompatible with "List[T@g]"
```
<img width="618" alt="image" src="https://user-images.githubusercontent.com/86978/163421436-db90092d-0fc3-4db1-9cc6-d8e995d599f0.png">

This is with Pylance v2022.4.1 .

*****

Some notes:

The specific `doc` decorator above could be rewritten so that it could be used as `@doc` instead of `@doc()`, and that does make the error go away. However, in my real use case, I want to be able to pass parameters to it, as in `@doc("hello")`.

The same thing also happens when I try to use the same typing strategy as Pandas' `@doc()` decorator
* https://github.com/pandas-dev/pandas/blob/dafa5dd/pandas/util/_decorators.py#L354
* https://github.com/pandas-dev/pandas/blob/dafa5dd/pandas/_typing.py#L150-L153

They do something like this, but it still results in error:
  
```python
FuncType = Callable[..., Any]
F = TypeVar("F", bound=FuncType)

def doc2() -> Callable[[F], F]:
    ...

@doc2()
def h(x: T, y: Union[T, List[T]]) -> None:
    pass

h(1, 2)       # `2` is underlined
h(1, [2, 3])  # OK
```

**Comment by erictraut** [comment]
Thanks for the bug report. This code was running afoul of a heuristic that I added to handle cases like this:

```python
def func(x: T | int) -> T | int: ...

reveal_type(func(3))
```

In this example, the type variable `T` is unbound because the argument `3` matches the `int` in the union. The return type is therefore `T | int`, but what is probably intended in this case is `int`. So there's a heuristic that removes unsolved TypeVars from unions in return types in certain cases.

I've added a bit more logic to this heuristic so it is not applied (incorrectly) in the code sample above.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.238, which I just published. It will also be included in the next release of pylance.
